README.md
m2pycrypto
==========

![Build Status](https://secure.travis-ci.org/carschar/m2pycrypto.png?branch=master)

This library is a drop-in replacement for the m2secret helper library.
m2secret is a handy wrapper around M2Crypto to make secure 256-bit AES
encryption very simple.

The m2pycrypto library has an identical interface and behavior to
m2secret, but with the underlying encryption library swapped out:
instead of M2Crypto it relies on the much more current PyCrypto.

The code is about 95% identical to m2secret and the interface is 100%
identical. Therefore, most of the credit goes to [Heikki Toivonen](http://www.heikkitoivonen.net/),
the original author of m2secret.


Why switch to PyCrypto?
=======================
* M2Crypto is unmaintained. No commits since January 2011.
* M2Crypto is a pain to compile on OS X and some other platforms.
* PyCrypto is the opposite: actively maintained and easy to install.
* PyCrypto supports Python 2.1 through 3.3.


How to replace m2secret with m2pycrypto
=======================================

1. Add 'm2pycrypto' to your environment with pip.
2. Replace "import m2secret" with "import m2pycrypto"
3. Run your tests
4. Verify your encrypted data

Yes, it should be that simple!


Installation
============
```
# Test and then install
python setup.py test
python setup.py install
```


Usage
=====
Here is simple usage to encrypt and decrypt.
```python
# Encrypt (auto-generate initialization vector and salt)
import m2pycrypto
secret = m2pycrypto.Secret()
ciphertext = secret.encrypt(plaintext, password)
# Better store these for later - you need them for decryption!
my_iv = secret.iv
my_salt = secret.salt


# Decrypt
import m2pycrypto
secret = m2secret.Secret(salt=my_salt, iv=my_iv)
plaintext = secret.decrypt(password)
```

More common use is encrypting before storing in a database, like 
for passwords. Instead of worrying about storing the 3 important
elements: ciphertext, salt, and iv, m2pycrypto has a serialize()
function which combines these into a single, ascii-safe string.

Similarly, there is a deserialize() function which re-parses that
specially-formatted string to extract the salt, iv, and ciphertext
which it needs to decrypt the data.

```python
# Encrypt (auto-generate initialization vector and salt)
import m2pycrypto
secret = m2pycrypto.Secret()
secret.encrypt(plaintext, password)
# This string can be saved to file or database
serialized_ciphertext = secret.serialize()


# Decrypt
import m2pycrypto
secret = m2secret.Secret()
secret.deserialize(serialized_ciphertext)
plaintext = secret.decrypt(password)
```


Links
=====
[Homepage of m2secret](http://www.heikkitoivonen.net/m2secret/)  
[Source code of m2secret (subversion)](http://svn.heikkitoivonen.net/svn/m2secret/trunk/)  
[PyCrypto Homepage](https://www.dlitz.net/software/pycrypto/)  
[PyCrypto on PyPI](https://pypi.python.org/pypi/pycrypto)  









HISTORY.md
0.0.1 / 2013-8-05
======

  * Meh.


tests/m2pycrypto_tests.py
from unittest import TestCase
import os
from binascii import hexlify

from mock import patch

from m2pycrypto import Secret
from m2pycrypto.m2pycrypto import pkcs7_decode, pkcs7_encode
from m2pycrypto.m2pycrypto import encrypt, decrypt


class TestSecretObject(TestCase):
    """Verify the behavior of Secret class."""

    def setUp(self):
        self.iv = 'iv_45678123456781234567812345678'
        self.salt = 'salt____123456781234567812345678'
        self.ciphertext = 'yadda yadda'
        self.iterations = 1000
        self.algorithm = 'aes_256_cbc'
        assert len(self.iv) == 32
        assert len(self.salt) == 32

    def TestParamsSaved(self):
        """Parameters sent to init() should be saved."""
        secret = Secret(
            iv=self.iv,
            salt=self.salt,
            ciphertext=self.ciphertext,
            iterations=self.iterations,
            algorithm=self.algorithm)
        self.assertEqual(secret.salt, self.salt)
        self.assertEqual(secret.iv, self.iv)
        self.assertEqual(secret.ciphertext, self.ciphertext)
        self.assertEqual(secret.iterations, self.iterations)
        self.assertEqual(secret.algorithm, self.algorithm)

    @patch.object(os, 'urandom')
    def TestDefaultParams(self, urandom):
        """Known defaults should be used when no parameters are given to init()."""
        urandom.return_value = 'x' * 32
        secret = Secret()
        self.assertEqual(len(secret.iv), 32)
        self.assertEqual(len(secret.salt), 32)
        urandom.assert_called_with(32)
        self.assertEqual(secret.ciphertext, None)
        self.assertEqual(secret.iterations, 1000)

    def TestUnicodePasswordBecomesUTF8Encrypt(self):
        """Encrypt should not blow up when given a unicode password."""
        secret = Secret(
            iv=self.iv,
            salt=self.salt,
            iterations=self.iterations,
            algorithm=self.algorithm)
        secret.encrypt('my cleartext', u'UNICODE PASSWORD!!')

    #def TestUnicodePasswordBecomesUTF8Decrypt(self):
    #    """Encrypt should not blow up when given a unicode password."""
    #    secret = Secret(
    #        iv=self.iv,
    #        salt=self.salt,
    #        iterations=self.iterations,
    #        algorithm=self.algorithm,
    #        ciphertext='')
    #    secret.decrypt(u'UNICODE PASSWORD!!')

    def TestSerialize(self):
        """Serialize should pack up the instance's iv, salt, and ciphertext."""
        hex_iv = hexlify(self.iv)
        hex_salt = hexlify(self.salt)
        hex_ciphertext = hexlify(self.ciphertext)
        test_serialized = "{0}|{1}|{2}".format(hex_iv, hex_salt, hex_ciphertext)
        secret = Secret(
            iv=self.iv,
            salt=self.salt,
            iterations=self.iterations,
            algorithm=self.algorithm,
            ciphertext=self.ciphertext)
        serialized = secret.serialize()
        self.assertEqual(serialized.count('|'), 2)
        self.assertEqual(serialized, test_serialized)


    def TestDeserialize(self):
        """Deserialize should save the unpacked iv, salt, and ciphertext onto the instance."""
        test_iv = 'i' * 32
        test_salt = 's' * 32
        test_ciphertext = 'test ciphertext'
        test_serialized = "{0}|{1}|{2}".format(
            hexlify(test_iv),
            hexlify(test_salt),
            hexlify(test_ciphertext))

        # Build a secret with values from setUp
        secret = Secret(
            iv=self.iv,
            salt=self.salt,
            iterations=self.iterations,
            algorithm=self.algorithm,
            ciphertext=self.ciphertext)

        secret.deserialize(test_serialized)

        self.assertEqual(secret.salt, test_salt)
        self.assertEqual(secret.iv, test_iv)
        self.assertEqual(secret.ciphertext, test_ciphertext)


class TestEncryptionFunctions(TestCase):
    def setUp(self):
        self.iv = 'i' * 32
        self.salt = 's' * 32
        self.key = 'k' * 32
        self.algorithm = 'aes_256_cbc'
        self.plaintext = 'Hello World!'
        self.ciphertext = ''

    #def test_normal_encrypt(self):
    #    ciphertext = encrypt(self.plaintext, self.key, self.iv, self.algorithm)
    #    print ciphertext
    #    self.assertEqual(ciphertext, self.ciphertext)
    #
    #def test_normal_decrypt(self):
    #    plaintext = encrypt(self.ciphertext, self.key, self.iv, self.algorithm)
    #    print plaintext


class TestPKDS7(TestCase):
    """Exercise the PKCS#7 Padding Functions."""

    def TestEncodeEmptyString(self):
        """Text of length zero should become a full pad block."""
        self.assertEqual(pkcs7_encode(''), '\x10' * 16)

    def TestEncodeFullString(self):
        """Text of length block size should get a full pad block added."""
        self.assertEqual(pkcs7_encode('d' * 16), 'd' * 16 + '\x10' * 16)

    def TestEncodeNormalStrings(self):
        """Text of length < block size should get padded to block size."""

        # Why not be exhaustive, especially if it helps someone
        # understand the algorithm?
        self.assertEqual(pkcs7_encode('a' * 1), 'a' * 1 + '\x0f' * 15)
        self.assertEqual(pkcs7_encode('a' * 2), 'a' * 2 + '\x0e' * 14)
        self.assertEqual(pkcs7_encode('a' * 3), 'a' * 3 + '\x0d' * 13)
        self.assertEqual(pkcs7_encode('a' * 4), 'a' * 4 + '\x0c' * 12)
        self.assertEqual(pkcs7_encode('a' * 5), 'a' * 5 + '\x0b' * 11)
        self.assertEqual(pkcs7_encode('a' * 6), 'a' * 6 + '\x0a' * 10)
        self.assertEqual(pkcs7_encode('a' * 7), 'a' * 7 + '\x09' * 9)
        self.assertEqual(pkcs7_encode('a' * 8), 'a' * 8 + '\x08' * 8)
        self.assertEqual(pkcs7_encode('a' * 9), 'a' * 9 + '\x07' * 7)
        self.assertEqual(pkcs7_encode('a' * 10), 'a' * 10 + '\x06' * 6)
        self.assertEqual(pkcs7_encode('a' * 11), 'a' * 11 + '\x05' * 5)
        self.assertEqual(pkcs7_encode('a' * 12), 'a' * 12 + '\x04' * 4)
        self.assertEqual(pkcs7_encode('a' * 13), 'a' * 13 + '\x03' * 3)
        self.assertEqual(pkcs7_encode('a' * 14), 'a' * 14 + '\x02' * 2)
        self.assertEqual(pkcs7_encode('a' * 15), 'a' * 15 + '\x01' * 1)

    def TestDecodeEmptyString(self):
        """An empty string should raise ValueError."""
        with self.assertRaises(ValueError):
            pkcs7_decode('')

    def TestDecodeBackToEmptyString(self):
        """A single full block of \x10 should decode to empty string."""
        self.assertEqual(pkcs7_decode('\x10' * 16), '')

    def TestDecodeBackToBlockSizedString(self):
        """A normal block followed by block of just padding should decode
        back to a string that was exactly the block size."""
        self.assertEqual(pkcs7_decode('d' * 16 + '\x10' * 16), 'd' * 16)





        # Yeah yeah, will write these soon.
        # <insert normal unit tests here>

        #############
        # Separate bank of tests that will prove the exact equivalence of
        # m2pycrypto and m2secret. These will only run if all of these packages
        # are installed:
        #   1. M2Crypto
        #   2. m2secret
        #   3. PyCrypto
        #   4. m2pycrypto

        # Add some logic to test imports and escape if not met

        # Tests go here

m2pycrypto/m2pycrypto.py
#!/usr/bin/env python
#
# Copyright (c) 2013 Carl Scharenberg <carl.scharenberg@gmail.com>
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

__all__ = ['SecretError', 'NoDataError', 'DecryptionError', 'EncryptionError',
           'PasswordError', 'Secret']

"""
Small utility and module for encrypting and decrypting data using symmetric-key
algorithms. By default uses 256-bit AES (Rijndael) using CBC, but some options
are configurable. PBKDF2 algorithm used to derive key from password.

Sample uses: passwords in INI files, password manager, encrypted files
"""

import os
import sys
import hmac
from hashlib import sha256
from binascii import hexlify, unhexlify

from Crypto.Cipher import AES
from Crypto.Protocol.KDF import PBKDF2


version = '0.1'
version_info = tuple([int(x) for x in version.split('.')])


class SecretError(Exception):
    """
    Base class for Secret-specific errors.
    """


class NoDataError(SecretError):
    """
    This exception will be raised if we don't have data to encrypt/decrypt.
    """


class DecryptionError(SecretError):
    """
    Failed to decrypt. Can happen with wrong password, for example.
    """


class EncryptionError(SecretError):
    """
    Failed to encrypt.
    """


class PasswordError(SecretError):
    """
    Problem with password(s).
    """


class Secret:
    """
    Stores a secret and has ways to decrypt the secret and set new secret.

    @warning: Once password is in memory, it will be possible
              to get the ciphertext as well. This may be possible
              even after the password is cleared due to Python memory
              management.
    @warning: If the password is used, secret will be decrypted and available
              in plain text in memory, possibly even after it has been
              explicitly cleared after use.
    @warning: If weak password is used, the
              encryption will not be of much help.
    """

    def __init__(self, iv=None, salt=None, ciphertext=None,
                 iterations=1000, algorithm='aes_256_cbc'):
        """
        Construct a Secret object.

        ciphertext, iv and salt can be None when originally created. The Secret
        is then considered to not hold any data. To set new data, call
        encrypt().

        @param iv: The IV, 256 bits (byte string 32 long)
        @param salt: The salt, 256 bits (byte string 32 long)
        @param ciphertext: The secret to hold
        @param iterations: The number of iterations to use with PBKDF2,
                           recommend 1000.
        @param param: The algorithm to use, recommend aes_256_cbc.
        """
        self.ciphertext = ciphertext
        self.iv = iv or os.urandom(32)
        self.salt = salt or os.urandom(32)
        self.iterations = iterations
        self.algorithm = algorithm

    def decrypt(self, password):
        """
        Decrypt.

        @param password: The password to decrypt data with.
        @return: Decrypted data
        """
        if not self.ciphertext or not self.iv or not self.salt or \
                        password is None:
            raise NoDataError

        # If the password is callable we'll assume it will return the
        # real password.
        try:
            password = password()
        except TypeError:
            pass

        # the crypto algorithms are unicode unfriendly
        if isinstance(password, unicode):
            password = password.encode('utf8')

        # derive 256 bit key using the pbkdf2 standard
        key = PBKDF2(password, self.salt, keylen=32, iterations=self.iterations)

        # Derive encryption key and HMAC key from it
        # See Practical Cryptography section 8.4.1.
        hmacKey = sha256(key + 'MAC').digest()
        encKey = sha256(key + 'encrypt').digest()
        del key

        # decrypt
        try:
            ret = decrypt(self.ciphertext, encKey, self.iv, self.algorithm)
        except Exception, e:
            # TODO Identify a more specific error to catch. For M2Crypto, was "EVP.EVPError"
            raise EncryptionError(str(e))
        finally:
            del encKey

        # Check MAC
        mac = ret[-64:]
        ret = ret[:-64]
        try:
            if hmac.new(hmacKey, ret + self.iv + self.salt,
                        sha256).hexdigest() != mac:
                raise DecryptionError('HMAC does not match')
        finally:
            del hmacKey

        return ret

    def encrypt(self, cleartext, password):
        """
        Encrypt.

        @param cleartext: The data to encrypt.
        @param password: The password to encrypt data with.
        @return: Encrypted data
        """
        if cleartext is None or password is None:
            raise NoDataError

        # If the password is callable we'll assume it will return the
        # real password.
        try:
            password = password()
        except TypeError:
            pass

        # the crypto algorithms are unicode unfriendly
        if isinstance(password, unicode):
            password = password.encode('utf8')

        # derive 256 bit encryption key using the pbkdf2 standard
        key = PBKDF2(password, self.salt, keylen=32, iterations=self.iterations)

        # Derive encryption key and HMAC key from it
        # See Practical Cryptography section 8.4.1.
        hmacKey = sha256(key + 'MAC').digest()
        encKey = sha256(key + 'encrypt').digest()
        del key

        # Add HMAC to cleartext so that we can check during decrypt if we got
        # the right cleartext back. We are doing sign-then-encrypt, which let's
        # us encrypt empty cleartext (otherwise we'd need to pad with some
        # string to encrypt). Practical Cryptography by Schneier & Ferguson
        # also recommends doing it in this order in section 8.2.
        mac = hmac.new(hmacKey,
                       cleartext + self.iv + self.salt,
                       sha256).hexdigest()
        del hmacKey

        # encrypt
        try:
            self.ciphertext = encrypt(cleartext + mac, encKey, self.iv,
                                      self.algorithm)
        except Exception, e:
            # TODO Identify a more specific error to catch. For M2Crypto, was "EVP.EVPError"
            raise EncryptionError(str(e))

        return self.ciphertext

    def serialize(self, serialize=None):
        """Serialize secret.

        @param serialize: None or callable that must accept string to serialize
        @return: Serialized string
        """
        if not self.ciphertext or not self.iv or not self.salt:
            raise NoDataError

        serialized = "%s|%s|%s" % (hexlify(self.iv), hexlify(self.salt),
                                   hexlify(self.ciphertext))
        if serialize is not None:
            serialize(serialized)
        return serialized

    def deserialize(self, deserialize):
        """Deserialize secret.

        @param deserialize: String or callable that must return the serialized form.
        """
        try:
            serialized = deserialize()
        except TypeError:
            serialized = deserialize

        iv, salt, ciphertext = serialized.split('|')
        self.iv, self.salt, self.ciphertext = unhexlify(iv), unhexlify(salt), unhexlify(ciphertext)

    def clear(self):
        try:
            del self.ciphertext
        except AttributeError:
            pass
        try:
            del self.iv
        except AttributeError:
            pass
        try:
            del self.salt
        except AttributeError:
            pass


def pkcs7_encode(text, k=16):
    n = k - (len(text) % k)
    return text + unhexlify(n * ("%02x" % n))


def pkcs7_decode(text, k=16):
    n = int(hexlify(text[-1]), 16)
    if n > k:
        raise ValueError("Input is not padded or padding is corrupt")
    return text[:-n]


def decrypt(ciphertext, key, iv, alg):
    """
    Decrypt ciphertext
    """
    assert len(key) == len(iv) == 32
    cipher = AES.new(key=key, mode=AES.MODE_CBC, IV=iv[:16])
    del key
    plaintext = pkcs7_decode(cipher.decrypt(ciphertext))
    return plaintext


def encrypt(plaintext, key, iv, alg):
    """
    Encrypt plaintext
    """
    assert len(key) == len(iv) == 32
    cipher = AES.new(key=key, mode=AES.MODE_CBC, IV=iv[:16])
    del key
    ciphertext = cipher.encrypt(pkcs7_encode(plaintext))
    assert ciphertext
    return ciphertext


def get_password(confirm=True):
    """get password"""
    import getpass

    password = getpass.getpass('password:')
    if confirm:
        if getpass.getpass('password (again):') != password:
            raise PasswordError('Passwords do not match')
    return password


def main():
    from optparse import OptionParser

    usage = 'usage: %prog [options]'
    parser = OptionParser(usage=usage,
                          version='%prog ' + version,
                          description='Encrypt or decrypt data with password using 256 bit AES (Rijndael) encryption in CBC mode. Key derived from password with PBKDF2 algorithm using 1000 iterations.')
    parser.add_option('-d', '--decrypt',
                      action='store_true', dest='decrypt', default=False,
                      help="Decryption mode.")
    parser.add_option('-e', '--encrypt',
                      action='store_true', dest='encrypt', default=False,
                      help="Encryption mode.")
    parser.add_option('-i', '--in',
                      dest='infile', metavar='INFILE',
                      help='INFILE to read in. Without this options reads stdin.')
    parser.add_option('-o', '--out',
                      dest='outfile', metavar='OUTFILE',
                      help="OUTFILE to output into. Without this option prints to stdout.")
    parser.add_option('-p', '--password',
                      metavar='PASSWORD', dest='password',
                      help="Supply PASSWORD from the command line (otherwise will be prompted for). Try to not use this option, since it is safer to be prompted for password.")

    (options, args) = parser.parse_args()
    if (options.decrypt and options.encrypt) or (not options.decrypt and not options.encrypt):
        parser.print_help()
        sys.exit(1)

    if not options.infile:
        options.args = ''.join(sys.stdin.readlines())
    else:
        options.args = ''

    if options.encrypt:
        secret = Secret()
        secret.encrypt(options.args or open(options.infile, 'rb').read(),
                       options.password or get_password)
        if options.outfile:
            f = open(options.outfile, 'wb')
            secret.serialize(f.write)
            f.close()
        else:
            secret.serialize(sys.stdout.write)
    else:
        secret = Secret()
        secret.deserialize((options.args or open(options.infile, 'rb').read()).strip())
        cleartext = secret.decrypt(options.password or (lambda: get_password(confirm=False)))
        if options.outfile:
            f = open(options.outfile, 'wb')
            f.write(cleartext)
            f.close()
        else:
            sys.stdout.write(cleartext)


if __name__ == "__main__":
    main()